storage: TestStoreRangeRemoveDead failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/d6adf24cae788d7cd967feadae8e9c0388ce5273

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestStoreRangeRemoveDead PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1080230&tab=buildLog

```
I190104 06:42:52.886487 62318 storage/client_raft_test.go:3108  SucceedsSoon: expected store 2 to be replaced; have [(n1,s1):1 (n3,s3):2 (n2,s2):3 (n4,s4):4 (n5,s5):5]
I190104 06:42:53.886810 62318 storage/client_raft_test.go:3108  SucceedsSoon: expected store 2 to be replaced; have [(n1,s1):1 (n3,s3):2 (n2,s2):3 (n4,s4):4 (n5,s5):5]
I190104 06:42:54.887227 62318 storage/client_raft_test.go:3108  SucceedsSoon: expected store 2 to be replaced; have [(n1,s1):1 (n3,s3):2 (n2,s2):3 (n4,s4):4 (n5,s5):5]
I190104 06:42:55.887578 62318 storage/client_raft_test.go:3108  SucceedsSoon: expected store 2 to be replaced; have [(n1,s1):1 (n3,s3):2 (n2,s2):3 (n4,s4):4 (n5,s5):5]
I190104 06:42:56.887958 62318 storage/client_raft_test.go:3108  SucceedsSoon: expected store 2 to be replaced; have [(n1,s1):1 (n3,s3):2 (n2,s2):3 (n4,s4):4 (n5,s5):5]
I190104 06:42:57.888461 62318 storage/client_raft_test.go:3108  SucceedsSoon: expected store 2 to be replaced; have [(n1,s1):1 (n3,s3):2 (n2,s2):3 (n4,s4):4 (n5,s5):5]
I190104 06:42:58.888828 62318 storage/client_raft_test.go:3108  SucceedsSoon: expected store 2 to be replaced; have [(n1,s1):1 (n3,s3):2 (n2,s2):3 (n4,s4):4 (n5,s5):5]
I190104 06:42:59.893397 62318 storage/store.go:1300  [s2] [n2,s2,r6/?:/Table/{SystemCon…-11}]: added to replica GC queue
I190104 06:42:59.895057 63776 util/stop/stopper.go:536  quiescing; tasks left:
3      storage.Replica: acquiring lease to gossip
I190104 06:42:59.895103 63775 util/stop/stopper.go:536  quiescing; tasks left:
1      storage.intentResolver: processing intents
1      [async] storage.replicate: processing replica
1      [async] kv.TxnCoordSender: heartbeat loop
I190104 06:42:59.895104 63776 util/stop/stopper.go:536  quiescing; tasks left:
2      storage.Replica: acquiring lease to gossip
W190104 06:42:59.895217 63483 storage/intent_resolver.go:756  [s1] failed to cleanup transaction intents: could not GC completed transaction anchored at /Local/Range/Table/SystemConfigSpan/Start/RangeDescriptor: result is ambiguous (server shutdown)
I190104 06:42:59.895243 63775 util/stop/stopper.go:536  quiescing; tasks left:
1      storage.intentResolver: processing intents
1      [async] storage.replicate: processing replica
I190104 06:42:59.895258 63775 util/stop/stopper.go:536  quiescing; tasks left:
1      [async] storage.replicate: processing replica
W190104 06:42:59.895369 63484 internal/client/txn.go:507  [replicate,s1,r6/1:/Table/{SystemCon…-11}] failure aborting transaction: node unavailable; try another peer; abort caused by: failed to send RPC: sending to all 1 replicas failed; last error: (err: node unavailable; try another peer) <nil>
E190104 06:42:59.895425 63484 storage/queue.go:844  [replicate,s1,r6/1:/Table/{SystemCon…-11}] change replicas of r6 failed: failed to send RPC: sending to all 1 replicas failed; last error: (err: node unavailable; try another peer) <nil>
E190104 06:42:59.895646 62999 storage/queue.go:844  [replicate,s1,r10/1:/Table/1{4-5}] [n1,s1,r10/1:/Table/1{4-5}]: unable to transfer lease to s5: [NotLeaseHolderError] r10: replica (n1,s1):1 not lease holder; lease holder unknown
W190104 06:42:59.896665 63116 gossip/gossip.go:1499  [n6] no incoming or outgoing connections
--- FAIL: TestStoreRangeRemoveDead (45.99s)
	client_raft_test.go:3108: condition failed to evaluate within 45s: expected store 2 to be replaced; have [(n1,s1):1 (n3,s3):2 (n2,s2):3 (n4,s4):4 (n5,s5):5]
		goroutine 62318 [running]:
		runtime/debug.Stack(0xa7a358200, 0xc420a5f4a0, 0x30694e0)
			/usr/local/go/src/runtime/debug/stack.go:24 +0xa7
		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x30c8f80, 0xc420518d20, 0xc420a5f470)
			/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:49 +0x10d
		github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeRemoveDead(0xc420518d20)
			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_raft_test.go:3108 +0x613
		testing.tRunner(0xc420518d20, 0x2be0f98)
			/usr/local/go/src/testing/testing.go:777 +0xd0
		created by testing.(*T).Run
			/usr/local/go/src/testing/testing.go:824 +0x2e0

```

SHA: https://github.com/cockroachdb/cockroach/commits/dadcaa493bca02996ebd87c0c85106e566893ce2

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestStoreRangeRemoveDead PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1084187&tab=buildLog

```
I190108 06:42:51.734257 63309 storage/store_snapshot.go:762  [replicate,s1,r9/1:/Table/1{3-4}] sending preemptive snapshot 50d37aba at applied index 16
I190108 06:42:51.734360 63309 storage/store_snapshot.go:805  [replicate,s1,r9/1:/Table/1{3-4}] streamed snapshot to (n2,s2):?: kv pairs: 7, log entries: 6, rate-limit: 8.0 MiB/sec, 0.00s
I190108 06:42:51.734599 63906 storage/replica_raftstorage.go:809  [s2,r9/?:{-}] applying preemptive snapshot at index 16 (id=50d37aba, encoded size=1196, 1 rocksdb batches, 6 log entries)
I190108 06:42:51.735290 63906 storage/replica_raftstorage.go:815  [s2,r9/?:/Table/1{3-4}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=0ms commit=0ms]
E190108 06:42:51.735580 63309 storage/queue.go:844  [replicate,s1,r9/1:/Table/1{3-4}] change replicas of r9 failed: node unavailable; try another peer
I190108 06:42:51.736150 63309 storage/store_snapshot.go:762  [replicate,s1,r7/1:/Table/1{1-2}] sending preemptive snapshot 885644b0 at applied index 16
I190108 06:42:51.736355 63309 storage/store_snapshot.go:805  [replicate,s1,r7/1:/Table/1{1-2}] streamed snapshot to (n3,s3):?: kv pairs: 7, log entries: 6, rate-limit: 8.0 MiB/sec, 0.00s
I190108 06:42:51.736643 63925 storage/replica_raftstorage.go:809  [s3,r7/?:{-}] applying preemptive snapshot at index 16 (id=885644b0, encoded size=1200, 1 rocksdb batches, 6 log entries)
I190108 06:42:51.737035 63925 storage/replica_raftstorage.go:815  [s3,r7/?:/Table/1{1-2}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
E190108 06:42:51.737281 63309 storage/queue.go:844  [replicate,s1,r7/1:/Table/1{1-2}] change replicas of r7 failed: node unavailable; try another peer
I190108 06:42:51.737731 63309 storage/store_snapshot.go:762  [replicate,s1,r18/1:/Table/2{2-3}] sending preemptive snapshot 6b7ccf13 at applied index 19
I190108 06:42:51.737826 63309 storage/store_snapshot.go:805  [replicate,s1,r18/1:/Table/2{2-3}] streamed snapshot to (n4,s4):?: kv pairs: 8, log entries: 9, rate-limit: 8.0 MiB/sec, 0.00s
I190108 06:42:51.738074 63928 storage/replica_raftstorage.go:809  [s4,r18/?:{-}] applying preemptive snapshot at index 19 (id=6b7ccf13, encoded size=1983, 1 rocksdb batches, 9 log entries)
W190108 06:42:51.738857 63928 storage/store.go:1646  [s4,r18/?:/Table/2{2-3}] unable to gossip on capacity change: node unavailable; try another peer
I190108 06:42:51.739374 63928 storage/replica_raftstorage.go:815  [s4,r18/?:/Table/2{2-3}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=1ms commit=0ms]
E190108 06:42:51.739647 63309 storage/queue.go:844  [replicate,s1,r18/1:/Table/2{2-3}] change replicas of r18 failed: node unavailable; try another peer
E190108 06:42:51.739787 63309 storage/queue.go:844  [replicate,s1,r1/1:/{Min-System/NodeL…}] [n1,s1,r1/1:/{Min-System/NodeL…}]: unable to transfer lease to s3: [NotLeaseHolderError] r1: replica (n1,s1):1 not lease holder; lease holder unknown
W190108 06:42:51.741649 62957 gossip/gossip.go:1499  [n5] no incoming or outgoing connections
W190108 06:42:51.741754 63301 gossip/gossip.go:1499  [n6] no incoming or outgoing connections
W190108 06:42:51.741808 63344 storage/raft_transport.go:583  while processing outgoing Raft queue to node 4: rpc error: code = Unavailable desc = transport is closing:
--- FAIL: TestStoreRangeRemoveDead (45.96s)
	client_raft_test.go:3108: condition failed to evaluate within 45s: expected store 2 to be replaced; have [(n1,s1):1 (n5,s5):2 (n2,s2):3 (n4,s4):4 (n3,s3):5]
		goroutine 62509 [running]:
		runtime/debug.Stack(0xa7a358200, 0xc421b3c030, 0x3096b60)
			/usr/local/go/src/runtime/debug/stack.go:24 +0xa7
		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x30f7120, 0xc42010f4a0, 0xc421b3c000)
			/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:49 +0x10d
		github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeRemoveDead(0xc42010f4a0)
			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_raft_test.go:3108 +0x613
		testing.tRunner(0xc42010f4a0, 0x2c08160)
			/usr/local/go/src/testing/testing.go:777 +0xd0
		created by testing.(*T).Run
			/usr/local/go/src/testing/testing.go:824 +0x2e0

```

--- a/pkg/storage/client_raft_test.go
+++ b/pkg/storage/client_raft_test.go
@@ -3012,7 +3012,12 @@ func TestStoreRangeRemoveDead(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	sc := storage.TestStoreConfig(nil)
 	sc.TestingKnobs.DisableReplicaRebalancing = true
-	mtc := &multiTestContext{storeConfig: &sc}
+	mtc := &multiTestContext{
+		storeConfig: &sc,
+		// This test needs to verify full replication, but as written the check
+		// only works if there's only one range initially.
+		startWithSingleRange: true,
+	}
 
 	storage.TimeUntilStoreDead.Override(&sc.Settings.SV, storage.TestTimeUntilStoreDead)
 	// Disable declined and failed reservations. Their default values are on
@@ -3066,13 +3071,26 @@ func TestStoreRangeRemoveDead(t *testing.T) {
 						panic(err)
 					}
 				}
+
+			case <-stopper.ShouldQuiesce():
+				return
+			}
+		}
+	})
+	stopper.RunWorker(ctx, func(ctx context.Context) {
+		tickerDur := storage.TestTimeUntilStoreDead / 2
+		ticker := time.NewTicker(tickerDur)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
 				// Force the repair queues on all alive stores to run.
 				for _, s := range nonDeadStores {
 					if err := s.ForceReplicationScanAndProcess(); err != nil {
 						panic(err)
 					}
 				}
-
 			case <-stopper.ShouldQuiesce():
 				return
 			}
@@ -3081,6 +3099,8 @@ func TestStoreRangeRemoveDead(t *testing.T) {
 
 	// Wait for up-replication, including at least one replica on s2 (deadStore).
 	testutils.SucceedsSoon(t, func() error {
+		// NB: this only checks the first range, so it'll give false positives
+		// if there's more than one (see the note at the beginning of the test).
 		replicas := getRangeMetadata(roachpb.RKeyMin, mtc, t).Replicas
 		var found bool
 		for _, replica := range replicas {

